ntoskrnl/ps/quota.c

```diff
@@ -143,45 +526,279 @@ PspInsertQuotaBlock(
+    }
+
+    DPRINT("PsReturnSharedPoolQuota(): Amount returned back (paged %lu -- non paged %lu)\n", AmountToReturnPaged, AmountToReturnNonPaged);
+
+    /*
+     * Decrement the process and reference counts.
+     * We might need to destroy the block if at
+     * that point we're dereferencing the last
+     * point where someone has used this block.
+     */
+    InterlockedDecrementSizeT(&QuotaBlock->ProcessCount);
+    if (InterlockedDecrementSizeT(&QuotaBlock->ReferenceCount) == 0)
+    {
+        PspDestroyQuotaBlock(QuotaBlock);
     }
 }
 
-/*
- * @implemented
+/**
+ * @brief
+ * Charges the shared (paged and non paged)
+ * pool quotas. The function is used exclusively
+ * by the Object Manager to manage quota charges
+ * handling of objects.
+ *
+ * @param[in] Process
+ * The process which quotas are to
+ * be charged within its quota block.
+ *
+ * @param[in] AmountToChargePaged
+ * The amount of paged quotas quotas to
+ * be charged.
+ *
+ * @param[in] AmountToChargeNonPaged
+ * The amount of non paged quotas to
+ * be charged.
+ *
+ * @return
+ * Returns the charged quota block, which it'll
+ * be used by the Object Manager to attach
+ * the charged quotas information to the object.
+ * If the function fails to charge quotas, NULL
+ * is returned to the caller.
+ */
+PEPROCESS_QUOTA_BLOCK
+NTAPI
+PsChargeSharedPoolQuota(
+    _In_ PEPROCESS Process,
+    _In_ SIZE_T AmountToChargePaged,
+    _In_ SIZE_T AmountToChargeNonPaged)
+{
+    NTSTATUS Status;
+
+    /* Sanity checks */
+    ASSERT(Process);
+    ASSERT(Process->QuotaBlock);
+
+    /* Do we have some paged pool quota to charge? */
+    if (AmountToChargePaged != 0)
+    {
+        /* We do, charge! */
+        Status = PspChargeProcessQuotaSpecifiedPool(NULL, Process->QuotaBlock, PsPagedPool, AmountToChargePaged);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("PsChargeSharedPoolQuota(): Failed to charge the shared pool quota (Status 0x%lx)\n", Status);
+            return NULL;
+        }
+    }
+
+    /* Do we have some non paged pool quota to charge? */
+    if (AmountToChargeNonPaged != 0)
+    {
+        /* We do, charge! */
+        Status = PspChargeProcessQuotaSpecifiedPool(NULL, Process->QuotaBlock, PsNonPagedPool, AmountToChargeNonPaged);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("PsChargeSharedPoolQuota(): Failed to charge the shared pool quota (Status 0x%lx). Attempting to return some paged pool back...\n", Status);
+            PspReturnProcessQuotaSpecifiedPool(NULL, Process->QuotaBlock, PsPagedPool, AmountToChargePaged);
+            return NULL;
+        }
+    }
+
+    /*
+     * As object quotas have been charged into the
+     * process' quota block, that block will gonna
+     * be used by the said process, so increment
+     * both process and reference counts.
+     */
+    InterlockedIncrementSizeT(&Process->QuotaBlock->ProcessCount);
```

Comment:
Should we really increase `ProcessCount` here? The reference count is needed in case the object outlives the process (we don't want to delete the block while an object still references it). But the process count seems to exist mostly for `PspReturnExcessQuotas`. It would make sense that if only objects, but not processes, are referencing a quota block, then the remaining quota would be considered excess (since the objects won't suddenly request additional quota).